### ISSUE TITLE
Slow initial loading of cloud-optimized GeoTIFF

### ISSUE THREAD

**Issue by LukasBommes** [openlayers/openlayers#16274]
**Describe the bug**
Given a large COG (for details see output of gdalinfo below), Openlayers downloads all header IFDs (up to 25 MB) before rendering the COG, causing very long initial page loads. You can observe this even on the [COG example page](https://openlayers.org/en/latest/examples/cog.html) where 2.1 MBs are downloaded (taking > 8 seconds on my end) before the COG is shown initially.

I confirmed that the entire header is loaded by comparing the header size of several COGs with the amount of data downloaded by the first GET requests to the COG. These values coincide closely. I obtained the header size via this [COG validation script](https://github.com/OSGeo/gdal/blob/master/swig/python/gdal-utils/osgeo_utils/samples/validate_cloud_optimized_geotiff.py)

**To Reproduce**
Take a large COG and serve it as shown in the [COG example](https://openlayers.org/en/latest/examples/cog.html). Observe the amount of data downloaded by the first requests in the networking tab of your browser's Developer tools.

**Expected behavior**
OpenLayers should only download the part of the header needed to render the current overview.

**Comment by M393** [comment]
In the example exactly 2MB are downloaded (+1Byte which may be a bug) which equals 8 Tiles a 256x256 * 4 byte per pixel.

What makes you think this is too much?

**Comment by LukasBommes** [comment]
Thanks a lot for your answer, @M393!

You are right. My conclusion that the issue occurs also in the COG example was drawn too quickly. The header size of the example COG is only 3622 Bytes. I striked out the sentence above.

However, this changes nothing about the problem that we experience with our own COGs where indeed the entire header is fetched before the COG is rendered. And since this header is for some COGs up to 25 MBs in size, this leads to very long initial page load times of up to 20 seconds.

**Comment by M393** [comment]
This is probably an issue with the [geotiff.js](https://github.com/geotiffjs/geotiff.js/issues) library, which is used by OpenLayers. Try reporting it there.

**Comment by LukasBommes** [comment]
Yes, I was thinking so, too. But then I found this sentence in the [geotiff.js docs](https://geotiffjs.github.io/geotiff.js/): "**All the metadata is parsed once the IFD is first parsed, thus the access to that is synchronous.**" So, I was wondering if geotiff.js parses the entire header first by design.

**Comment by ahocevar** [comment]
Can we agree that this is not an OpenLayers issue?  Snd as a contributor to geotiff.js, I'm inclined to say that it isn't a geotiff.js issue either.

The fact that a COG is a cloud optimized *format* does not mean that the data provider took care to optimize the data properly for the intended use (web application in this case). If the header of a single GeoTIFF gets too big, then maybe using a [COG pyramid](https://openlayers.org/en/latest/examples/cog-pyramid.html) instead of a single COG is the way to go.

**Comment by mzmcgrat** [comment]
Many thanks to the commenters on this issue.  I was experiencing something similar, but now I realize that the COG format won't save me.  Using a COG Pyramid would seem to re-introduce the drawbacks I was trying to avoid by using a single COG with overviews in the first place (and the problem would remain when loading the highest zoom level).

Since my use case is write once, read many times for visualization in a web app, I can encode my data into 1 or 3 channels of PNG tiles and package them into a raster PMtiles file, which gives me the low-overhead and serverless results I was seeking originally.  Perhaps something similar might be an option for the issue you're facing. Cheers!

**Comment by LukasBommes** [comment]
Thanks for the hint, @mzmcgrat! I have indeed move on to PMTiles already and it works like a dream for my use case.

**Comment by ricardodepinho** [comment]
[COG is an OGC official format](https://www.ogc.org/announcement/cloud-optimized-geotiff-cog-published-as-official-ogc-standard/), so it would be great for the users and for the Openlayers and Geotiff.js projects if they could be 100% compliant to it.

I've got this same issue using a Portuguese nationwide COG file to create a web map with ol.geotiff to view COG, as they were made to be used, in HTTP-Range streaming.
https://github.com/openlayers/openlayers/discussions/17030

I've been using QGIS in HTTP-Range to access the same COG file with amazing speed and no problems.

So I would appreciate both projects, OL and GeoTIFF.js (@LukasBommes @ahocevar) to consider this issue a bug and proceed to solve it ASAP.

Thank you in advance!
Best regards.
